### DISCUSSION TITLE
Wix Installer is tries to download .Net 8 when already installed

### DISCUSSION THREAD

**Discussion by cornerbowlsoftware** [wixtoolset/issues#8733] [Questions]
I have customers in STIGed environments where the installer is failing because Wix is trying to download .Net from an air gapped machine when it's already installed.  I have run tests with air gapped machines and cannot reproduce the issue.  While working with another customer on a corporate non-STIGed machine, I noticed the following lines:

```
[129C:2330][2024-09-04T10:31:11]i338: Acquiring package: Net8Redist, payload: Net8Redist, download from: https://download.visualstudio.microsoft.com/download/pr/3980ab0a-379f-44a0-9be6-eaf74c07a3b3/bd1cc6107ff3d8fe0104d30f01339b74/dotnet-runtime-8.0.7-win-x64.exe
[4A3C:5BA4][2024-09-04T10:31:12]i305: Verified acquired payload: Net8Redist at path: C:\ProgramData\Package Cache\.unverified\Net8Redist, moving to: C:\ProgramData\Package Cache\E016B47FF8F179F1EDF641CCA20130E04951509F0F69F564F290BFB0AD81D7DB\dotnet-runtime-8.0.7-win-x64.exe.
```

 However just above the log files states: 

```
[129C:5A20][2024-09-04T10:31:05]i052: Condition 'DOT_NET_VER >= v8.0.0' evaluates to true.
[129C:5A20][2024-09-04T10:31:05]i101: Detected package: Net8Redist, state: Present, cached: No, install registration state: (permanent), cache registration state: (permanent)
```

Why is the installer downloading files it doesn't need?

**Comment by cornerbowlsoftware** [comment]
My DetectCondition is:

DetectCondition="DOT_NET_VER &gt;= v8.0.0" 

Is this backwards?  I reviewed my previous Wix3 installers and they always ignore the .Net installation when the detect condition result equals true.

**Comment by cornerbowlsoftware** [comment]
I customer sent screenshots in PDF format of the logs and I can verify Wix is trying to install .Net 8 over and over again.

**Comment by csdahlberg** [comment]
I saw something similar when using `NetFx462RedistAsPrereq`. Our installer was correctly detecting that a newer .NET Framework version was already installed, but was attempting to download the 4.6.2 installer solely to cache it locally (not to actually install it). Switching to `NetFx462WebAsPrereq` avoided the attempt to download the 4.6.2 installer when 4.6.2+ was already installed.

As a short-term, no-code-change-required workaround, I also found that manually copying the .NET Framework 4.6.2 installer to `%folder_with_our_installer%\redist\ndp462-kb3151800-x86-x64-allos-enu.exe` allowed `NetFx462RedistAsPrereq` to succeed without reaching out to the internet.

Might either of those help you?

**Comment by cornerbowlsoftware** [comment]
@csdahlberg  I am using the following .wxs to detect .Net 8. Are you suggesting I set a parameter such as PrereqPackage="no"?

Here is my wxs:
```
<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs" xmlns:netfx="http://wixtoolset.org/schemas/v4/wxs/netfx">

    <Fragment>        
        <netfx:DotNetCoreSearch RuntimeType="desktop" Platform="x64" MajorVersion="8" Variable="DOT_NET_VER" />

        <PackageGroup Id="Net8DesktopRedist">
            <ExePackage 
				Id="Net8DesktopRedist" 
				DisplayName="Microsoft .NET 8.0 Desktop Runtime (v8.0.8)" 
				Cache="keep" 
				PerMachine="yes" 
				Permanent="yes" 
				Vital="yes" 
				DetectCondition="DOT_NET_VER &gt;= v8.0.0" 
				InstallArguments="/norestart /quiet">
                <ExePackagePayload 
					Description="Microsoft .NET 8.0 Desktop Runtime (v8.0.8) - Windows x64" 
					ProductName="Microsoft .NET Desktop Runtime - 8.0.8 (x64)" 					
					DownloadUrl="https://download.visualstudio.microsoft.com/download/pr/907765b0-2bf8-494e-93aa-5ef9553c5d68/a9308dc010617e6716c0e6abd53b05ce/windowsdesktop-runtime-8.0.8-win-x64.exe"
					Hash="27484FFB1E9CE5E6290CDA8F5F49563CF4A9E2692AA57429FCD0D3DE4F30FC2FCE204B1B120349ED50712E95D3FA51037EBAF9B7FD60C41856857B1372E0EAC7"
					Name="windowsdesktop-runtime-8.0.8-win-x64.exe"
					Size="58492592"
					Version="8.0.8.33916"
					/>
            </ExePackage>
        </PackageGroup>
    </Fragment>
</Wix>
```

**Reply by csdahlberg** [reply]
I never did confirm exactly what was causing our failures with `NetFx462RedistAsPrereq`, but comparing your wxs to https://github.com/wixtoolset/wix/blob/main/src/ext/NetFx/wixlib/NetFx462.wxs, I wonder if using `Cache="remove"` for your `ExePackage` might do it.

**Comment by cornerbowlsoftware** [comment]
@csdahlberg I looked at the docs and they state:

remove: Always uncache the package after Cache, Install, Modify, Repair, Uninstall, and Layout.

Which sounds like it's only being used after the cache.  I just created an embedded installer that includes the full .net runtime and verified it doesn't try to download anything so I think I'm good now.
